sdks/java/io/csv/src/test/java/org/apache/beam/sdk/io/csv/CsvIOWriteTest.java

```diff
@@ -0,0 +1,534 @@
+                            .withFieldValue("badfield", Arrays.asList("a", "b"))
+                            .withFieldValue("ok", "abcdefg")
+                            .build()))
+                .setRowSchema(arraySchema)
+                .apply(CsvIO.writeRows().withCSVFormat(CSVFormat.DEFAULT).to(to)));
+
+    assertThrows(
+        "schema should not contain row field",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(
+                    Create.of(
+                        Row.withSchema(rowSchema)
+                            .withFieldValue("badfield", rows.get(0))
+                            .withFieldValue("ok", "abcdefg")
+                            .build()))
+                .setRowSchema(rowSchema)
+                .apply(CsvIO.writeRows().to(to)));
+  }
+
+  @Test
+  public void invalidSchemaFields() {
+    String to = tmpFolder.getRoot().getAbsolutePath() + "/somewhere";
+
+    assertThrows(
+        "schema field names should not be empty",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(
+                    CsvIO.writeRows()
+                        .to(to)
+                        .withSchemaFields(Arrays.asList("", "string", "anInt"))));
+
+    assertThrows(
+        "schema fields should not be an empty list",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(CsvIO.writeRows().to(to).withSchemaFields(Collections.emptyList())));
+
+    assertThrows(
+        "field names should match schema",
+        IllegalArgumentException.class,
+        () ->
+            errorPipeline
+                .apply(Create.of(rows))
+                .setRowSchema(ALL_DATA_TYPES_SCHEMA)
+                .apply(
+                    CsvIO.writeRows()
+                        .to(to)
+                        .withSchemaFields(Arrays.asList("doesnotexist", "string", "anInt"))));
+  }
+
+  @Test
+  public void withCompression() {
+    String to = "with_compression";
+    String regex = String.format("^%s.*$", to);
+    pipeline
+        .apply(Create.of(rows).withRowSchema(ALL_DATA_TYPES_SCHEMA))
+        .apply(
+            CsvIO.writeRows()
+                .to(tmpFolder.getRoot().getAbsolutePath() + "/" + to)
+                .withCompression(Compression.GZIP));
+    pipeline.run().waitUntilFinish();
+    String[] files = filesMatching(regex);
+    for (String name : files) {
+      Path p = Paths.get(tmpFolder.getRoot().getAbsolutePath(), name);
+      File f = new File(p.toString());
+      assertTrue(Compression.GZIP.isCompressed(f.getAbsolutePath()));
+    }
+  }
+
+  @Test
+  public void withInvalidType() {
+    assertThrows(
+        IllegalArgumentException.class,
```

Comment:
Can you match against the error string as well to ensure we are catching the exception from the right place and not just any `IllegalArgumentException`